platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,395 @@
+
+No accessibility checking is performed on the content-only deploys. That means that content writers don't have a proactive way to learn if their content includes accessibility errors. Adding accessibility checking to the CMS is on the roadmap, but it doesn't currently exist.
+
+Additionally, the current invalid content reporting is inconvenient. Broken links are extracted from the build log as a CSV. And accessibility errors are manually communicated from the Frontend Tools team.
+
+##### VA.gov Users
+
+To enable a faster publishing workflow, all tests (except link checking) are intentionally skipped during content-only deploys. That creates an opportunity for end users to encounter accessibility errors. Additionally, due to inconsistencies between how the CMS handles redirects and how `vets-website` handles redirects, it's possible for users to encounter broken links in production.
+
+##### FE Tools team
+
+Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with the CMS team.
+
+##### Anybody that is trying to ship code for the VA.gov front end
+
+If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.
+
+## High Level Design
+
+Once the [build separation work](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2719) is complete, there will be separate content and application builds. These distinct builds will prevent failures in one from blocking deployments to the other.
+
+The Metalsmith static content build will be extracted to the `content-build` repo and deleted from the `vets-website` repo and deployment pipeline. Webpack will handle the application build for the `vets-website` repo, and Metalsmith will handle the static content build for the `content-build` repo.
+
+The broken link validation will continue using the existing the `check-broken-link` Metalsmith plugin and `glean-broken-links` script. Since broken links currently block content-only deploys, then we will continue blocking content-only deploys when broken links are detected.
+
+The validation of accessibility errors will take place in a scheduled job that runs hourly during workdays, instead of happening as part of the full deploy. Since the accessibility check relies on the output of the content build, the scheduled job will run a script to produce a build that only has the steps necessary for accessibility checking. That build will only be used for reporting accessibility errors, and will have no impact on deploys. Those reports will be delivered to the #cms-team channel via existing Slack integrations.
```

Comment:
I saw above that there's a plan to get a11y testing into the CMS, maybe it's worth accelerating that so the site doesn't risk falling out of section 508/WCAG AA compliance? I think code reviews should mitigate it from the apps end, but content I could see tripping up quite a bit - maybe if it's just trying to recommend a Drupal plugin or something (cc @meganhkelley)

Reply:
Accessibility checks for content on VA.gov are split over a variety of places — part of the CMS, part of automated tests on vets-website, and as manual reviews done before and after products and features launch. Features within the CMS itself would fall under purview of the CMS team — but we currently maintain the CMS build pipeline, which has some accessibility checks, which is why it came up in this design doc. 

In other words, let's chat more!

Reply:
The issue is most of the content from the CMS is in fields that lack the bigger html wrappers, so it can not be accurately tested since it is not rendered the same way it is by the  metalsmith build.  

There are some WYSIWIG instances in the cms but those are fairly restricted as far as what html tags people can use and what they can add.  It is not a guarantee that what comes out of them is accessible, but it is often the case that what may be needed for accessiblity may not be attainable in the wysiwig.    There is some scanning we could do and should do, but it will not be the panacea we want it to be.

Reply:
we decided that since the a11y tests take 30mins to run, hourly didn't makes sense, so we're backing this off to daily